File: java/src/main/java/io/github/xesam/unilink/seal/resolver/PolicyProcessor.java
package io.github.xesam.unilink.seal.resolver;

import io.github.xesam.unilink.seal.LinkSealErrorCode;
import io.github.xesam.unilink.seal.ResolutionException;
import io.github.xesam.unilink.seal.model.MissingPolicy;
import io.github.xesam.unilink.seal.model.PolicyConfig;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class PolicyProcessor {
  private static final Pattern TEMPLATE_VAR_PATTERN = Pattern.compile("\\{([^}]+)\\}");

  public PolicyResult process(Map<String, String> variables, PolicyConfig config) {
    Map<String, String> defaults = config.defaults();
    Map<String, String> result = new LinkedHashMap<>();
    List<String> missing = new ArrayList<>();

    for (Map.Entry<String, String> entry : variables.entrySet()) {
      String key = entry.getKey();
      String value = entry.getValue();
      if (value != null) {
        result.put(key, value);
      } else if (config.missing() == MissingPolicy.DEFAULT && defaults.containsKey(key)) {
        result.put(key, defaults.get(key));
      } else {
        missing.add(key);
      }
    }

    return new PolicyResult(result, missing);
  }

  public Map<String, String> applyMissingPolicy(
      String template, Map<String, String> variables, PolicyConfig config) {
    PolicyResult processed = process(variables, config);

    if (!processed.missing().isEmpty()) {
      if (config.missing() == MissingPolicy.ERROR) {
        throw new ResolutionException(
            LinkSealErrorCode.MISSING_VARIABLE,
            "Missing required variables: " + String.join(", ", processed.missing()));
      }

      if (config.missing() == MissingPolicy.IGNORE) {
        // Check if any missing variable is in a path segment
        Set<String> pathSegmentVars = extractPathSegmentVariables(template);
        List<String> missingInPath = new ArrayList<>();
        for (String missingVar : processed.missing()) {
          if (pathSegmentVars.contains(missingVar)) {
            missingInPath.add(missingVar);
          }
        }

        if (!missingInPath.isEmpty()) {
          throw new ResolutionException(
              LinkSealErrorCode.INVALID_TEMPLATE,
              "Policy \"ignore\" is unsafe for path-segment variables ["
                  + String.join(", ", missingInPath)
                  + "]. Use \"error\" or \"default\" policy for path variables.");
        }
      }
    }

    return processed.result();
  }

  /**
   * Extracts variable names that appear in path segments (not in query parts).
   * Path segment variables are those in {var} expressions before the '?' query separator
   * and not using query operators {?...} or {&...}.
   */
  private Set<String> extractPathSegmentVariables(String template) {
    Set<String> pathVars = new HashSet<>();

    // Split at '?' to get path part only
    String[] parts = template.split("\\?", 2);
    String pathPart = parts[0];

    Matcher matcher = TEMPLATE_VAR_PATTERN.matcher(pathPart);
    while (matcher.find()) {
      String expression = matcher.group(1);

      // Skip query operators
      if (expression.startsWith("?") || expression.startsWith("&")) {
        continue;
      }

      // Handle comma-separated variable lists (e.g., {x,y,z})
      String[] varNames = expression.split(",");
      for (String varName : varNames) {
        pathVars.add(varName.trim());
      }
    }

    return pathVars;
  }

  public record PolicyResult(Map<String, String> result, List<String> missing) {}
}

File: javascript/packages/unilink-seal-resolver/src/PolicyProcessor.ts
import type { PolicyConfig } from 'unilink-seal';
import { ResolutionError } from './errors.js';

export class PolicyProcessor {
  process(
    variables: Record<string, string | null>,
    config: PolicyConfig
  ): { result: Record<string, string>; missing: string[] } {
    const defaults = config.defaults ?? {};
    const result: Record<string, string> = {};
    const missing: string[] = [];

    for (const [key, value] of Object.entries(variables)) {
      if (value !== null && value !== undefined) {
        result[key] = value;
      } else if (config.missing === 'default' && defaults[key] !== undefined) {
        result[key] = defaults[key];
      } else {
        missing.push(key);
      }
    }

    return { result, missing };
  }

  applyMissingPolicy(
    template: string,
    variables: Record<string, string | null>,
    config: PolicyConfig
  ): { result: Record<string, string>; error?: { code: string; message: string } } {
    const { result, missing } = this.process(variables, config);

    if (missing.length > 0) {
      if (config.missing === 'error') {
        return {
          result,
          error: {
            code: 'MISSING_VARIABLE',
            message: `Missing required variables: ${missing.join(', ')}`,
          },
        };
      }

      if (config.missing === 'ignore') {
        // Check if any missing variable is in a path segment
        const pathSegmentVars = this.extractPathSegmentVariables(template);
        const missingInPath = missing.filter(v => pathSegmentVars.has(v));

        if (missingInPath.length > 0) {
          return {
            result,
            error: {
              code: 'INVALID_TEMPLATE',
              message: `Policy "ignore" is unsafe for path-segment variables [${missingInPath.join(', ')}]. Use "error" or "default" policy for path variables.`,
            },
          };
        }
      }
    }

    return { result };
  }

  /**
   * Extracts variable names that appear in path segments (not in query parts).
   * Path segment variables are those in {var} expressions before the '?' query separator
   * and not using query operators {?...} or {&...}.
   */
  private extractPathSegmentVariables(template: string): Set<string> {
    const pathVars = new Set<string>();

    // Split at '?' to get path part only
    const [pathPart] = template.split('?');

    const varPattern = /\{([^}]+)\}/g;
    let match: RegExpExecArray | null;

    while ((match = varPattern.exec(pathPart)) !== null) {
      const expression = match[1];

      // Skip query operators
      if (expression.startsWith('?') || expression.startsWith('&')) {
        continue;
      }

      // Handle comma-separated variable lists (e.g., {x,y,z})
      const varNames = expression.split(',');
      for (const varName of varNames) {
        pathVars.add(varName.trim());
      }
    }

    return pathVars;
  }
}
